overlord/devicestate/devicestate.go

```diff
@@ -1049,6 +1136,53 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 	return tss, nil
 }
 
+func verifyModelValidationSets(st *state.State, newModel *asserts.Model, offline bool, deviceCtx snapstate.DeviceContext) (*snapasserts.ValidationSets, error) {
+	vSets, err := assertstate.ValidationSetsFromModel(st, newModel, assertstate.ValidationSetsModelOptions{
+		Offline: offline,
+	}, deviceCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := checkForInvalidSnapsInModel(newModel, vSets); err != nil {
+		return nil, err
+	}
+
+	if err := checkForRequiredSnapsNotInModel(newModel, vSets); err != nil {
+		return nil, err
+	}
+
+	return vSets, nil
+}
+
+func checkForRequiredSnapsNotInModel(model *asserts.Model, vSets *snapasserts.ValidationSets) error {
+	snapsInModel := make(map[string]bool, len(model.RequiredWithEssentialSnaps()))
+	for _, sn := range model.RequiredWithEssentialSnaps() {
+		snapsInModel[sn.SnapName()] = true
+	}
+
+	for _, sn := range vSets.RequiredSnaps() {
+		if !snapsInModel[sn] {
+			return fmt.Errorf("cannot have required snap in validation set that is not present in the model: %s", sn)
```

Comment:
Nitpick: feels simpler and reads nicer (even nicer if you could include which vs)
```suggestion
			return fmt.Errorf("missing required snap %q in model", sn)
```